src/mo_frontend/typing.ml

```diff
@@ -1048,11 +1051,49 @@ and infer_exp'' env exp : T.typ =
       else env
     in
     infer_obj env' obj_sort.it dec_fields exp.at
-  | ObjE exp_fields ->
+  | ObjE (exp_fields, exp_bases) ->
+    let open List in
     check_ids env "object" "field"
-      (List.map (fun (ef : exp_field) -> ef.it.id) exp_fields);
-    let fts = List.map (infer_exp_field env) exp_fields in
-    T.Obj (T.Object, List.sort T.compare_field fts)
+      (map (fun (ef : exp_field) -> ef.it.id) exp_fields);
+    let fts = map (infer_exp_field env) exp_fields in
+    let fls = map (fun {T.lab; _} -> lab) fts in
+    let bases = map (fun b -> infer_exp env b, b) exp_bases in
+
+    (* removing explicit fields from the bases *)
+    let strip (base_t, base) =
+      let s, tfs =
+        try T.as_obj base_t with Invalid_argument _ ->
+          error env base.at "M0093"
+            "expected object type, but expression produces type%a"
+            display_typ_expand base_t in
+      (* forbid actors as bases *)
+      if s = T.Actor then
+        error env base.at "M0178"
+          "actors cannot serve as bases in record extensions";
+      T.(Obj (Object, filter (fun f -> not (mem f.lab fls)) tfs)) in
+    let stripped_bases = map strip bases in
+
+    (* label disjointness of stripped bases *)
+    let rec disjoint = function
+      | [] | [_] -> ()
+      | (h, h_exp) :: t ->
+        let avoid {T.lab; _} =
+          let avoid_labels b bls =
+            if mem lab bls then
+              begin
+                local_error env b.at "M0177"
+                  "ambiguous field in base%a"
+                  display_lab lab;
+                info env h_exp.at "field also present in base, here (consider overwriting)"
+              end in
+          iter (fun (b_t, b) -> avoid_labels b (map (fun {T.lab; _} -> lab) (T.as_obj b_t |> snd))) t in
+        iter avoid (T.as_obj h |> snd);
+        disjoint t in
+    disjoint (map2 (fun b_t b -> b_t, b) stripped_bases exp_bases);
+
+    (* TODO: var in stripped_bases? *)
+    let t_base = T.(fold_left glb (Obj (Object, [])) stripped_bases) in
+    T.(glb t_base (Obj (Object, sort T.compare_field fts)))
```

Comment:
So that I get a sorted result :-)

Reply:
Ah ok, its because you remember the types of the bases as object types, not list of fields, so need to merge them using glb. Ok.
The other option would be to just remember the list of fields taken from each object, flatten, append tfs, and sort.

Reply:
Right